File: frontend/lib/api/portfolio.ts
/**
 * Portfolio API services.
 */

import { api } from './client';

export interface PortfolioHolding {
    id: string;
    ticker: string;
    shares: number | null;
    average_cost: number | null;
    added_at: string;
    notes: string | null;
}

// Type alias for cleaner imports
export type Holding = PortfolioHolding;

export interface Portfolio {
    id: string;
    name: string;
    description: string | null;
    created_at: string;
    updated_at: string;
    holdings: PortfolioHolding[];
}

export interface PortfolioListResponse {
    portfolios: Portfolio[];
    total: number;
}

export interface CreatePortfolioData {
    name: string;
    description?: string;
}

export interface UpdatePortfolioData {
    name?: string;
    description?: string;
}

export interface AddHoldingData {
    ticker: string;
    shares?: number;
    average_cost?: number;
    notes?: string;
}

export interface UpdateHoldingData {
    shares?: number;
    average_cost?: number;
    notes?: string;
}

/**
 * Get all portfolios for the current user.
 */
export async function getPortfolios(): Promise<PortfolioListResponse> {
    return api.get<PortfolioListResponse>('/api/portfolios');
}

/**
 * Get a specific portfolio by ID.
 */
export async function getPortfolio(id: string): Promise<Portfolio> {
    return api.get<Portfolio>(`/api/portfolios/${id}`);
}

/**
 * Create a new portfolio.
 */
export async function createPortfolio(data: CreatePortfolioData): Promise<Portfolio> {
    return api.post<Portfolio>('/api/portfolios', data);
}

/**
 * Update a portfolio.
 */
export async function updatePortfolio(id: string, data: UpdatePortfolioData): Promise<Portfolio> {
    return api.put<Portfolio>(`/api/portfolios/${id}`, data);
}

/**
 * Delete a portfolio.
 */
export async function deletePortfolio(id: string): Promise<void> {
    return api.delete(`/api/portfolios/${id}`);
}

/**
 * Add a stock to a portfolio.
 */
export async function addStockToPortfolio(
    portfolioId: string,
    data: AddHoldingData
): Promise<PortfolioHolding> {
    return api.post<PortfolioHolding>(`/api/portfolios/${portfolioId}/stocks`, data);
}

/**
 * Update a stock in a portfolio.
 */
export async function updateStockInPortfolio(
    portfolioId: string,
    ticker: string,
    data: UpdateHoldingData
): Promise<PortfolioHolding> {
    return api.put<PortfolioHolding>(`/api/portfolios/${portfolioId}/stocks/${ticker}`, data);
}

/**
 * Remove a stock from a portfolio.
 */
export async function removeStockFromPortfolio(
    portfolioId: string,
    ticker: string
): Promise<void> {
    return api.delete(`/api/portfolios/${portfolioId}/stocks/${ticker}`);
}

File: backend/app/models/portfolio.py
"""
Portfolio and PortfolioHolding models for user investment tracking.
"""

import uuid
from datetime import datetime
from decimal import Decimal
from sqlalchemy import String, Text, DateTime, ForeignKey, Numeric, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from typing import TYPE_CHECKING

from ..database import Base

if TYPE_CHECKING:
    from .user import User


class Portfolio(Base):
    """
    Portfolio model representing a user's investment portfolio.
    
    Attributes:
        id: Unique identifier (UUID)
        user_id: Reference to the owning user
        name: Portfolio name
        description: Optional description
        created_at: Timestamp of creation
        updated_at: Timestamp of last update
    """
    
    __tablename__ = "portfolios"
    
    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )
    description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )
    
    # Relationships
    user: Mapped["User"] = relationship(
        "User",
        back_populates="portfolios",
    )
    holdings: Mapped[list["PortfolioHolding"]] = relationship(
        "PortfolioHolding",
        back_populates="portfolio",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    
    def __repr__(self) -> str:
        return f"<Portfolio(id={self.id}, name={self.name}, user_id={self.user_id})>"


class PortfolioHolding(Base):
    """
    PortfolioHolding model representing a stock position within a portfolio.
    
    Attributes:
        id: Unique identifier (UUID)
        portfolio_id: Reference to the parent portfolio
        ticker: Stock ticker symbol
        shares: Number of shares held
        average_cost: Average cost per share
        added_at: When the position was first added
        notes: Optional notes about the position
    """
    
    __tablename__ = "portfolio_holdings"
    __table_args__ = (
        UniqueConstraint("portfolio_id", "ticker", name="uq_portfolio_ticker"),
    )
    
    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )
    portfolio_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("portfolios.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    ticker: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        index=True,
    )
    shares: Mapped[Decimal | None] = mapped_column(
        Numeric(18, 8),
        nullable=True,
    )
    average_cost: Mapped[Decimal | None] = mapped_column(
        Numeric(18, 4),
        nullable=True,
    )
    added_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    notes: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )
    
    # Relationships
    portfolio: Mapped["Portfolio"] = relationship(
        "Portfolio",
        back_populates="holdings",
    )
    
    def __repr__(self) -> str:
        return f"<PortfolioHolding(id={self.id}, ticker={self.ticker}, shares={self.shares})>"
